docs/InstrumentationEvents.md

```diff
@@ -25,74 +25,29 @@ Instrumentation Event:
+|2      | CONNECT             |                                                                                                                                                                                                               | Consumer connected with broker |
+|3      | GROUP_JOIN          | {`groupId`, `memberId`, `leaderId`, `isLeader`, `memberAssignment`, `groupProtocol`, `duration`}                                                                                                                   | Consumer has joined the group |
+|4      | FETCH_START         | {}                                                                                                                                                                                                            | Starting to fetch messages from broker |
+|5      | FETCH               | {`numberOfBatches`, `duration`}                                                                                                                                                                           | Fetched messages from the broker |
+|6      | START_BATCH_PROCESS | {`topic`, `partition`, `highWatermark`, `offsetLag`, `offsetLagLow`, `batchSize`, `firstOffset`, `lastOffset`}                                                                                | Starting to process the batch |
+|7      | END_BATCH_PROCESS   | {`topic`, `partition`, `highWatermark`, `offsetLag`, `offsetLagLow`, `batchSize`, `firstOffset`, `lastOffset`, `duration`}                                                                  | Processed the batch |
+|8      | COMMIT_OFFSETS      | {`groupId`, `memberId`, `groupGenerationId`, `topics`}                                                                                                                                                | Committed resolved offsets |
+|9     | STOP                |                                                                                                                                                                                                               | Consumer has stopped |
+|10     | DISCONNECT          |                                                                                                                                                                                                               | Consumer has disconnected |
+|-      | CRASH               | {`error`, `groupId`}                                                                                                                                                                                      | Consumer has crashed. After this, consumer will try to perform a full restart if possible, otherwise it will stay hanging. Ideally you should listen to this event and handle yourself. |
```

Comment:
> otherwise it will stay hanging

This needs to be clarified. There's nothing in KafkaJS that sits around after a crash. If the consumer crashes with a non-retriable error, it will stop and the process will exit unless there's some other (non-KafkaJS) work pending in the event loop. For example (here I removed the special case for KafkaJSNumberOfRetriesExceeded to simulate a fatal error):

```
$ node examples/consumer.js
{"level":"INFO","timestamp":"2020-06-08T06:31:41.554Z","logger":"kafkajs","message":"[Consumer] Starting","groupId":"test-group"}
{"level":"INFO","timestamp":"2020-06-08T06:31:41.625Z","logger":"kafkajs","message":"[Runner] Consumer has joined the group","groupId":"test-group","memberId":"example-consumer-ac9b4fe8-83b3-49e2-ba2b-c6d545cfc61d","leaderId":"example-consumer-ac9b4fe8-83b3-49e2-ba2b-c6d545cfc61d","isLeader":true,"memberAssignment":{"topic-test":[0]},"groupProtocol":"RoundRobinAssigner","duration":69}
{"level":"ERROR","timestamp":"2020-06-08T06:31:41.673Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"topic-test","partition":0,"offset":"0","stack":"Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"ERROR","timestamp":"2020-06-08T06:31:42.023Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"topic-test","partition":0,"offset":"0","stack":"Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"ERROR","timestamp":"2020-06-08T06:31:42.740Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"topic-test","partition":0,"offset":"0","stack":"Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"ERROR","timestamp":"2020-06-08T06:31:44.368Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"topic-test","partition":0,"offset":"0","stack":"Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"ERROR","timestamp":"2020-06-08T06:31:47.116Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"topic-test","partition":0,"offset":"0","stack":"Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"ERROR","timestamp":"2020-06-08T06:31:53.614Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"topic-test","partition":0,"offset":"0","stack":"Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"ERROR","timestamp":"2020-06-08T06:31:53.614Z","logger":"kafkajs","message":"[Consumer] Crash: KafkaJSNumberOfRetriesExceeded: 💣","groupId":"test-group","retryCount":5,"stack":"KafkaJSNonRetriableError\n  Caused by: Error: 💣\n    at Runner.eachMessage (/Users/tommybrunn/workspace/kafkajs/examples/consumer.js:35:13)\n    at Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:173:20)\n    at onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:307:20)\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:356:21\n    at invoke (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:51:7)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:53\n    at new Promise (<anonymous>)\n    at /Users/tommybrunn/workspace/kafkajs/src/utils/concurrency.js:60:20\n    at /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:346:11"}
{"level":"INFO","timestamp":"2020-06-08T06:31:53.621Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"test-group"}
```

Reply:
@Nevon You are right.
But the service will not know unless it's listening to this event. Right?

Reply:
If you're embedding KafkaJS into some other kind of service that keeps the process running, then yes, you would need to listen to the crash event to know that the consumer has stopped.

Reply:
Do you think this would be helpful
```
Note: If your service is doing some other tasks and also consuming messages.
In the case of CRASH, the consumer process will try to restart if possible 
(depend on the type of error) otherwise the consumer will stop and exit,
and the service will not know. You should listen to the `CRASH` event if you 
want your service to react.
```

Reply:
I would put it like this:

> In the case of `CRASH`, the consumer will try to restart itself in the case of retriable errors. If the error is not retriable, the consumer will stop and exit. If your application wants to react to the error, such as cleanly shutting down resources, restarting the consumer itself, or exiting the process entirely, it should listen to the `CRASH` event.

I don't like to make too many assumptions about the environment that people are using KafkaJS in. If they are just using KafkaJS itself, the process would exit when the consumer stops, so there wouldn't be any reason to listen to the crash event. It's only when the consumer is embedded into some other application (which is certainly not uncommon) that this can be necessary.